update default upstream when forking repo during PR creation

daviddl9 opened the pull request:
This PR updates the behaviour to set the default upstream when forking a repo during PR creation. It ensures that the behaviour when performing a fork during PR creation remains consistent with `gh repo fork`.

Closes #10277

This bash script can be used to verify these changes: 
```bash
#!/bin/bash
set -e

# Configuration
UPSTREAM_REPO="BagToad/testing-public-1"
GITHUB_HOST="github.com"
GH_BIN="/Users/davidlivingston/Work/cli/bin/gh" # update with your own path

# Function to cleanup on exit or error
cleanup() {
    local exit_code=$?
    echo "Performing cleanup..."
    
    # Clean up local directory
    if [ -n "$TEST_DIR" ] && [ -d "$TEST_DIR" ]; then
        echo "Removing local test directory..."
        rm -rf "$TEST_DIR"
    fi

    # Delete fork if it exists
    if [ -n "$FORK_EXISTS" ]; then
        echo "Deleting forked repository..."
        if "$GH_BIN" repo delete "$FORK_OWNER/$FORK_NAME" --confirm; then
            echo "Successfully deleted fork"
        else
            echo "Failed to delete fork"
            exit 1
        fi
    fi

    exit $exit_code
}

# Set up cleanup trap
trap cleanup EXIT

# Get the authenticated user
FORK_OWNER=$("$GH_BIN" api user -q .login)
FORK_NAME="testing-public-1"
echo "Authenticated as: $FORK_OWNER"

# Check if fork already exists
if "$GH_BIN" repo view "$FORK_OWNER/$FORK_NAME" &>/dev/null; then
    echo "Fork already exists at $FORK_OWNER/$FORK_NAME"
    echo "Please delete it first or use a different test account"
    exit 1
fi

# Test setup
echo "Setting up test environment..."

# Create a new directory for testing
TEST_DIR=$(mktemp -d)
cd "$TEST_DIR"
echo "Working in temporary directory: $TEST_DIR"

# Clone the CLI repository
echo "Cloning github.com/$UPSTREAM_REPO..."
git clone "https://$GITHUB_HOST/$UPSTREAM_REPO.git"
cd testing-public-1

# Create a test branch
TEST_BRANCH="test-default-repo-$(date +%s)"
git checkout -b "$TEST_BRANCH"

# Make a small change
echo "Making test changes..."
DATE_COMMENT="// Test change made on $(date)"
echo "$DATE_COMMENT" >> README.md

git add README.md
git commit -m "test: Add date comment for testing default repo setting"

# Create PR using gh CLI
echo "Creating pull request..."
GH_DEBUG=api "$GH_BIN" pr create \
    --title "test: Verify default repo setting" \
    --body "This is a test PR to verify that the default repository is set correctly when creating PRs from forks."

# Mark that fork exists for cleanup
FORK_EXISTS=1

TEST_BRANCH="test-default-repo-$(date +%s)-2"
git checkout -b "$TEST_BRANCH"

# Make a small change
echo "Making test changes..."
DATE_COMMENT="// Test change made on $(date)"
echo "$DATE_COMMENT" >> README.md

git add README.md
git commit -m "test: Add date comment for testing default repo setting"

# Create PR using ghCLI
echo "Creating pull request..."
GH_DEBUG=api "$GH_BIN" pr create \
    --title "test: Verify default repo setting 2" \
    --body "This is a test PR 2 to verify that the default repository is set correctly when creating PRs from forks."

echo "Verifying default repository setting..."

# First list all remotes and configs for debugging
echo "Current remotes:"
git remote -v

# Now check the gh-resolved setting
DEFAULT_REPO=$(git config --get remote.upstream.gh-resolved || echo "not set")
echo "Default repository setting: $DEFAULT_REPO"

if [ "$DEFAULT_REPO" = "base" ]; then
    echo "✅ Success: Default repository was set correctly"
else
    echo "❌ Error: Default repository was not set correctly. Upstream remote exists but gh-resolved setting is missing."
    exit 1
fi

echo "Test completed successfully"
```

daviddl9 committed:
update default upstream when forking repo

update default upstream when forking repo

remove unnecessary variable

daviddl9 committed:
Merge branch 'trunk' into update-pr-create

andyfeller commented:
Thanks for opening this PR and contributing to the GitHub CLI, @daviddl9!

I'm starting to review these changes while doing a bit of manual testing to understand the new experience.  Will leave my notes and review shortly.

andyfeller commented:
## Manual testing

### `v2.66.1` experience

```shell
tinyfists/gh-pr-test ‹cli-10277*›$ git add .
tinyfists/gh-pr-test ‹cli-10277*›$ git commit
[cli-10277 822a8da] Add file for testing cli/cli#10277
 1 file changed, 3 insertions(+)
 create mode 100644 cli-10277.md


tinyfists/gh-pr-test ‹cli-10277›$ gh pr create
? Where should we push the 'cli-10277' branch? Create a fork of tinyfists/gh-pr-test

Creating pull request for andyfeller:cli-10277 into main in tinyfists/gh-pr-test

? Title (required) Add file for testing cli/cli#10277
? Body <Received>
? What's next? Submit
Changed tinyfists/gh-pr-test remote to "upstream"
Added andyfeller/gh-pr-test as remote "origin"
remote: Repository not found.
fatal: repository 'https://github.com/andyfeller/gh-pr-test.git/' not found
waiting 2 seconds before retrying...
remote: 
remote: 
To https://github.com/andyfeller/gh-pr-test.git
 * [new branch]      HEAD -> cli-10277
branch 'cli-10277' set up to track 'origin/cli-10277'.
https://github.com/tinyfists/gh-pr-test/pull/7


tinyfists/gh-pr-test ‹cli-10277›$ cat .git/config 
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true
[remote "upstream"]
	url = https://github.com/tinyfists/gh-pr-test.git
	fetch = +refs/heads/*:refs/remotes/upstream/*
[branch "main"]
	remote = upstream
	merge = refs/heads/main
[remote "origin"]
	url = https://github.com/andyfeller/gh-pr-test.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "cli-10277"]
	remote = origin
	merge = refs/heads/cli-10277
```

### PR experience

```shell
tinyfists/gh-testing-01 ‹cli-10277*›$ git add .
tinyfists/gh-testing-01 ‹cli-10277*›$ git commit
[cli-10277 e804616] Add file to test for cli/cli#10458
 1 file changed, 3 insertions(+)
 create mode 100644 cli-10277.md


tinyfists/gh-testing-01 ‹cli-10277›$ .../bin/gh pr create
? Where should we push the 'cli-10277' branch? Create a fork of tinyfists/gh-testing-01

Creating pull request for andyfeller:cli-10277 into main in tinyfists/gh-testing-01

? Title (required) Add file to test for cli/cli#10458
? Body <Received>
? What's next? Submit
Changed tinyfists/gh-testing-01 remote to "upstream"
Added andyfeller/gh-testing-01 as remote "origin"
! Repository andyfeller/gh-testing-01 set as the default repository. To learn more about the default repository, run: gh repo set-default --help
remote: Repository not found.
fatal: repository 'https://github.com/andyfeller/gh-testing-01.git/' not found
waiting 2 seconds before retrying...
remote: 
remote: 
To https://github.com/andyfeller/gh-testing-01.git
 * [new branch]      HEAD -> cli-10277
branch 'cli-10277' set up to track 'origin/cli-10277'.
https://github.com/tinyfists/gh-testing-01/pull/1


tinyfists/gh-testing-01 ‹cli-10277›$ cat .git/config
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true
[remote "upstream"]
	url = https://github.com/tinyfists/gh-testing-01.git
	fetch = +refs/heads/*:refs/remotes/upstream/*
	gh-resolved = base
[branch "main"]
	remote = upstream
	merge = refs/heads/main
[remote "origin"]
	url = https://github.com/andyfeller/gh-testing-01.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "cli-10277"]
	remote = origin
	merge = refs/heads/cli-10277
```

jtmcg commented:
Popping in, here: that Bash script looks like a good candidate for an acceptance test around this.

andyfeller commented on pkg/cmd/pr/create/create.go:
```diff
@@ -1002,6 +1003,15 @@ func handlePush(opts CreateOptions, ctx CreateContext) error {
+				stderr := opts.IO.ErrOut
+				fmt.Fprintf(stderr, "%s Repository %s set as the default repository. To learn more about the default repository, run: gh repo set-default --help\n", cs.WarningIcon(), cs.Bold(ghrepo.FullName(headRepo)))
+			}
+		}
```
## Suggestions

Out of all of following idle thoughts, I think only a handful are reasonable to knock out here now versus waiting for new issues being raised:

1. Remove `stdin` requirement for messaging
2. Use local variable versus hardcoding remote name
3. Return the error raised by setting default remote

## Notes

Some of my idle thoughts reviewing this logic:

1. Why does `stdin` being TTY affect whether we output this message?
2. How does such conditional logic apply to the `Added ... as remote` message?
3. When do we use a variable for the remote name versus hardcoding this everywhere?
4. How does this differ from other places in code where we handle this?
5. Does this logic need to ensure no other remote is marked as default before setting `upstream`?

### TTY affecting messaging

It seems like `gh repo fork` care about `stdin`, `stdout`, `stderr`... 

https://github.com/cli/cli/blob/ae2f4b7c5b49dd1b5855a8c2d283e0b02fb636eb/pkg/cmd/repo/fork/fork.go#L187

https://github.com/cli/cli/blob/ae2f4b7c5b49dd1b5855a8c2d283e0b02fb636eb/pkg/cmd/repo/fork/fork.go#L380-L383

whereas `gh repo clone` just cares about `stdout` being TTY:

https://github.com/cli/cli/blob/ae2f4b7c5b49dd1b5855a8c2d283e0b02fb636eb/pkg/cmd/repo/clone/clone.go#L217-L221

Definitely want to revisit where different types or classes of messages are displayed so we had a little more consistency.  In the meantime, I assume `gh pr create` would follow the same behavior as `gh repo clone` without factoring in `stdin` if we even bother to check.


### Setting and unsetting

My main concern with the changes as they stand is comparing this versus `gh repo set-default` which checks `func (r Remotes) ResolvedRemote() (*Remote, error)` to determine if it should unset a remote before setting a new one 🤔 

https://github.com/cli/cli/blob/ae2f4b7c5b49dd1b5855a8c2d283e0b02fb636eb/pkg/cmd/repo/setdefault/setdefault.go#L228-L242

It feels there is an edge case that can emerge around `upstream` remote separately of the server-side repository forking the command does.  However, let's see about users reporting errors before dealing with that concern.

### Other places where we set default repository

Just because I know this messaging is in a few different places, I always look at how consistent our behavior is.

https://github.com/cli/cli/blob/ae2f4b7c5b49dd1b5855a8c2d283e0b02fb636eb/pkg/cmd/repo/clone/clone.go#L209-L221

https://github.com/cli/cli/blob/ae2f4b7c5b49dd1b5855a8c2d283e0b02fb636eb/pkg/cmd/repo/fork/fork.go#L372-L383

andyfeller committed:
Suggestions to enhance upstream default handling

This commit brings this code more into alignment with other places:

1. Using a local variable for upstream remote name instead of hardcoding everywhere
2. Raising error raised by setting default remote
3. Removing condition for displaying upstream default message
4. Bring message logic more inline with other places within function

andyfeller reviewed (approved):
Looks good, @daviddl9! ✨ 

In an effort to keep the train rolling, I commit a few enhancements based on the suggestions below.  Some of the other idle thoughts are things we can revisit in the future.

andyfeller commented:
> Popping in, here: that Bash script looks like a good candidate for an acceptance test around this.

@cli/code-reviewers : Perhaps its worth discussing at our next retro about when we write out `acceptance` tests for OSS contributions; I suspect there's a bit of tension between lowering the barrier for contributors versus ensuring user experience works as expected.

daviddl9 commented:
Hey @andyfeller thanks for taking the time to look through the PR! Mostly agree with the ideas that you had in mind. Had a little suggestion about how we handle the case when setting the default upstream fails: 

```
if didForkRepo {
			err := gitClient.SetRemoteResolution(context.Background(), upstreamName, "base")
			if err != nil {
				return fmt.Errorf("error setting upstream as default: %w", err)
			}
```

The core operation here is creating a PR. Setting the default upstream is a secondary, convenience operation that helps with future workflows but isn't critical to the PR's existence or functionality. Therefore, I'm thinking a warning would be more appropriate here since:
- The PR can still function perfectly without this setting
- The user can manually set the upstream later if needed
- We probably shouldn't break a primary workflow due to a non-critical configuration issue

What do you think? Maybe something along the lines of: 

```
if didForkRepo {
    if err := gitClient.SetRemoteResolution(context.Background(), upstreamName, "base"); err != nil {
        // Log warning instead of failing
        cs := opts.IO.ColorScheme()
        fmt.Fprintf(opts.IO.ErrOut, "%s Warning: Could not set upstream as default: %v\n", cs.WarningIcon(), err)
        fmt.Fprintf(opts.IO.ErrOut, "You can set it manually later using: gh repo set-default\n")
    } 
    :
}
```

andyfeller commented:
> The core operation here is creating a PR. Setting the default upstream is a secondary, convenience operation that helps with future workflows but isn't critical to the PR's existence or functionality. Therefore, I'm thinking a warning would be more appropriate here since:
> 
> * The PR can still function perfectly without this setting
> * The user can manually set the upstream later if needed
> * We probably shouldn't break a primary workflow due to a non-critical configuration issue
> 
> What do you think? Maybe something along the lines of:
> 
> ```
> if didForkRepo {
>     if err := gitClient.SetRemoteResolution(context.Background(), upstreamName, "base"); err != nil {
>         // Log warning instead of failing
>         cs := opts.IO.ColorScheme()
>         fmt.Fprintf(opts.IO.ErrOut, "%s Warning: Could not set upstream as default: %v\n", cs.WarningIcon(), err)
>         fmt.Fprintf(opts.IO.ErrOut, "You can set it manually later using: gh repo set-default\n")
>     } 
>     :
> }
> ```

🤔 Fair concern; let's see some of the other places we're doing this:

https://github.com/cli/cli/blob/8365e0b3bbdecbf4be9246585ca69c458ad92b85/pkg/cmd/config/set/set.go#L62-L67

https://github.com/cli/cli/blob/8365e0b3bbdecbf4be9246585ca69c458ad92b85/pkg/cmd/repo/rename/rename.go#L159-L168

@daviddl9 : what is the likelihood this will fail leaving this in a bad state?

daviddl9 commented:
@andyfeller thanks for looking into this and for the prompt response! fair question - yeah realistically although the likelihood of failure is small, in the rare event that it does fail it could lead to a confusing / frustrating user experience where the core PR creation fails because of this. If you're ok, I could put up another small PR to change this?

andyfeller commented:
> @andyfeller thanks for looking into this and for the prompt response! fair question - yeah realistically although the likelihood of failure is small, in the rare event that it does fail it could lead to a confusing / frustrating user experience where the core PR creation fails because of this. If you're ok, I could put up another small PR to change this?

@daviddl9 : I'm going to bring this back into the issue so we can discuss it with the user who reported this as well as my fellow maintainer for discussion.  It isn't clear cut whether this should be a hard or soft error; I personally think rather that this should be a hard error.  Generally, we've adopted a _"let's wait for this to emerge as a problem before we fix"_ but let's bubble this up there.

_UPDATE: See https://github.com/cli/cli/issues/10277#issuecomment-2675234537_
